KEYCLOAK-11019 Initial support for lazy offline user-session loading

martin-kanis opened the pull request:
This PR includes #7722 + rebase to current master + adds model tests

thomasdarimont committed:
KEYCLOAK-11019 InfinispanUserSessionProvider should fallback to database query if in-memory lookup fails for offline sessions

Allow to configure pre-loading of offline user-sessions

We now allow to configure whether offline user-sessions should be
preloaded on startup from the database.

We now trigger offline user-sessions loading when
UserSessionProvider#getOfflineUserSessionsStream(realm,user) is called.

This ensures that the user can revoke the offline_access consent
via the account console. Additionally the (offline) user consents
are now displayed correctly for a user in the admin-console.

Offline user-sessions are now correctly display if offline user-sessions
are loaded lazily.

Add note about unsupported lazy offline session loading by broker info:
Currently we don't support lazy offline user-session lookup by
brokerSessionId and brokerUserId.

See: https://github.com/keycloak/keycloak/pull/7722#pullrequestreview-578669985

Keep behavior of 'logout all sessions' and don't delete dangling offline sessions.
As discussed in https://github.com/keycloak/keycloak/pull/7722#discussion_r566989120

Throw on unsupported dynamic offline user-session lookups from database

thomasdarimont committed:
KEYCLOAK-11019 Fix exception message for unsupported brokered offline user-session lookups

thomasdarimont committed:
KEYCLOAK-11019 Fix naming in UserSessionPredicate

thomasdarimont committed:
KEYCLOAK-11019 Revise queries for offline user-session loading

Removed redundant condition checking.

We use the `coalesce` function to make the queries work for a
specific realm / client or all, e.g.:
```
clientSess.clientId = coalesce(CAST(:clientId as text), clientSess.clientId)
```
Note that the explicit cast is required to treat null values as text to avoid
confusing databases.

thomasdarimont committed:
KEYCLOAK-11019 Add indices to speed up lazy offline user-session lookups

thomasdarimont committed:
KEYCLOAK-11019 Revise indices according to analysis from code review

martin-kanis committed:
KEYCLOAK-11019 Add model tests + fix broken cluster tests + move migration changelog to 14

vramik reviewed (changes requested):
Thank you for the PR. I went thru the original PR and there has been done huge amount of work. Thanks again.

Few notes:

- There was some issues in GHA 
    - model tests were cancelled after 329m. But I can see it passed in @martin-kanis's fork so I assume it is not related. 
    - cluster tests are failing
- I haven't found any documentation PR, is there a plan to send one?
- the changeSet should be updated, please see inline.

Otherwise it looks good to me.

hmlnarik commented:
> model tests were cancelled after 329m. But I can see it passed in @martin-kanis's fork so I assume it is not related.

That's correct, it is an independent intermittent issue, tracked as https://issues.redhat.com/browse/KEYCLOAK-18091

martin-kanis commented on model/jpa/src/main/resources/META-INF/jpa-changelog-13.0.0.xml:
```diff
@@ -87,4 +87,23 @@
         <renameColumn tableName="REALM_ATTRIBUTE" oldColumnName="VALUE_NEW" newColumnName="VALUE" columnDataType="NCLOB"/>
     </changeSet>
 
+    <changeSet author="keycloak" id="13.0.0-KEYCLOAK-11019">
```
Thanks, moved.

martin-kanis commented:
The failing cluster tests were caused by a syntax error in SQL query on MySQL. The `coalesce` + `cast` functions didn't work as expected at least on MySQL in clustered environment with provided `null` values. I removed the `coalesce` from all queries because from the keycloak code we always called the queries with `null` values. 
I also removed some newly introduced methods which weren't called from anywhere. 

```
WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (ServerService Thread Pool -- 69) SQL Error: 1064, SQLState: 42000
2021-04-02T22:46:06.7960979Z AuthenticationSessionClusterTest ++ 22:45:45,473 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (ServerService Thread Pool -- 69) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'longtext), persistent0_.REALM_ID)' at line 1
2021-04-02T22:46:06.7965147Z AuthenticationSessionClusterTest ++ 22:45:45,484 FATAL [org.keycloak.services] (ServerService Thread Pool -- 69) Error during startup: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
2021-04-02T22:46:06.7969007Z AuthenticationSessionClusterTest ++ 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
2021-04-02T22:46:06.7972268Z AuthenticationSessionClusterTest ++ 	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1575)
2021-04-02T22:46:06.7976018Z AuthenticationSessionClusterTest ++ 	at org.hibernate.query.internal.AbstractProducedQuery.getSingleResult(AbstractProducedQuery.java:1614)
2021-04-02T22:46:06.7980927Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.jpa.session.JpaUserSessionPersisterProvider.getUserSessionsCount(JpaUserSessionPersisterProvider.java:460)
2021-04-02T22:46:06.7986423Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.session.UserSessionPersisterProvider.getUserSessionsCount(UserSessionPersisterProvider.java:133)
2021-04-02T22:46:06.7993299Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.computeLoaderContext(OfflinePersistentUserSessionLoader.java:67)
2021-04-02T22:46:06.8001438Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.computeLoaderContext(OfflinePersistentUserSessionLoader.java:37)
2021-04-02T22:46:06.8007571Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer$2.run(InfinispanCacheInitializer.java:92)
```

hmlnarik commented:
https://master-jenkins/job/universal-test-pipeline-server/406/

martin-kanis commented:
https://master-jenkins/job/universal-test-pipeline-server/438/

vramik reviewed (commented):
Thanks @martin-kanis! 

Th PR looks good to me. From my POV there is just documentation PR missing and we've discussed with @hmlnarik that after those changes in queries the index analysis (which was done with the original PR) should re-checked.

hmlnarik commented:
Indices checked, they look good. @vramik , is the docs the only remaining concern?

vramik reviewed (approved):
@hmlnarik yes, the docs, are the only remaining concern, thank you for checking the indices.

hmlnarik reviewed (approved):
Pipelines ran successfully. Thank you for the contribution, @thomasdarimont and @martin-kanis !
